presto/presto.sh

```diff
@@ -48,9 +48,35 @@ connector.name=hive-hadoop2
 hive.metastore.uri=thrift://localhost:9083
 EOF
 
+# Compute memory settings based on Spark's settings.
+# We use "tail -n 1" since overrides are applied just by order of appearance.
+SPARK_EXECUTOR_MB=$(grep spark.executor.memory /etc/spark/conf/spark-defaults.conf | tail -n 1 | cut -d '=' -f 2 | cut -d 'm' -f 1)
+SPARK_EXECUTOR_CORES=$(grep spark.executor.cores /etc/spark/conf/spark-defaults.conf | tail -n 1 | cut -d '=' -f 2)
+SPARK_EXECUTOR_OVERHEAD_MB=$(grep spark.yarn.executor.memoryOverhead /etc/spark/conf/spark-defaults.conf | tail -n 1 | cut -d '=' -f 2)
+SPARK_EXECUTOR_COUNT=$(( $(nproc) / ${SPARK_EXECUTOR_CORES} ))
+
+# Add up overhead and allocated executor MB for container size.
+SPARK_CONTAINER_MB=$(( ${SPARK_EXECUTOR_MB} + ${SPARK_EXECUTOR_OVERHEAD_MB} ))
+PRESTO_JVM_MB=$(( ${SPARK_CONTAINER_MB} * ${SPARK_EXECUTOR_COUNT} ))
+
+# Give query.max-memorr-per-node 60% of Xmx; this more-or-less assumes a
+# single-tenant use case rather than trying to allow many concurrent queries
+# against a shared cluster.
+# Subtract out SPARK_EXECUTOR_OVERHEAD_MB in both the query MB and reserved
+# system MB as a crude approximation of other unaccounted overhead that we need
+# to leave betweenused bytes and Xmx bytes. Rounding down by integer division
+# here also effectively places round-down bytes in the "general" pool.
+PRESTO_QUERY_NODE_MB=$(( ${PRESTO_JVM_MB} * 6 / 10 - ${SPARK_EXECUTOR_OVERHEAD_MB} ))
+PRESTO_RESERVED_SYSTEM_MB=$(( ${PRESTO_JVM_MB} * 4 / 10 - ${SPARK_EXECUTOR_OVERHEAD_MB} ))
+
+# Compute the distributed max-memory by multiplying worker count by the
+# per-worker memory allocations.
+WORKER_COUNT=$(/usr/share/google/get_metadata_value attributes/dataproc-worker-count)
+PRESTO_QUERY_DISTIBUTED_MB=$(( ${WORKER_COUNT} * ${PRESTO_QUERY_NODE_MB} ))
```

Comment:
Do you know if this is necessary? It breaks with scaling, which is why we set the YARN equivalent to 10 TB